crates/runtime/src/memory.rs

```diff
@@ -484,12 +488,119 @@ impl SharedMemory {
     pub fn vmmemory_ptr(&self) -> *const VMMemoryDefinition {
         &self.0.read().unwrap().def.0 as *const _
     }
+
+    /// Implementation of `memory.atomic.notify` for this shared memory.
+    pub fn atomic_notify(&self, addr_index: u64, count: u32) -> Result<u32, TrapCode> {
+        let definition = &self.0.read().unwrap().def.0;
+        definition.validate_addr(addr_index, 4, 4)?;
+        // SAFETY: checked `addr_index` above
+        Ok(unsafe { self.unchecked_atomic_notify(addr_index, count) })
+    }
+
+    /// Unchecked implementation of `memory.atomic.notify` for this shared memory.
+    ///
+    /// # Safety
+    /// The caller must ensure that `addr_index` is a valid address in the memory.
+    pub unsafe fn unchecked_atomic_notify(&self, addr_index: u64, count: u32) -> u32 {
```

Comment:
One thing I might recommend, could the `unsafe` functions here be removed? That way the libcalls to delegate validation to these functions and that would remove the need to have `unsafe` interfaces entirely ideally.

Reply:
So, the spectests expect the alignment check before any other trap and even in the non-shared  case.
Because I can't get to a `SharedMemory` object in the non-shared case, the `validate_addr` has to live elsewhere, so I moved it in the fast path to  `VMMemoryDefinition`.

And in the shared memory case, I have already validated the Atomic address, so I thought, I create an unchecked version called with the already validated address to not do the same work twice.